docs/src/adr/014adr-precise-records.md

```diff
@@ -0,0 +1,670 @@
+           \cup [type : {"2a"}, bal : Ballot, val : Value]
+           \cup [type : {"2b"}, acc : Acceptor, bal : Ballot, val : Value]
+```           
+
+## 4. Options
+
+There are several solutions to the issue of precise record typing.
+
+### 4.1. Support only plain records
+
+In this case, we would only allow to mix records that have exactly the same
+shape. As a result, when we use variants, we would have to add extra
+fields to all of them. This solution is not very different from the current
+implementation of record type checking, though it would allow us to quickly
+detect spelling errors.
+
+**Blocker.** This does not look like a real solution, as it would immediately
+render the existing examples invalid. Moreover, they would be no obvious way to
+repair these examples.
+
+### 4.2. Support plain records and variants, but no row typing
+
+In this scenario, the type checker would issue an error, if a record expression
+accesses a field that is outside of its type:
+
+```tla
+FieldAccess ==
+  LET m == [ a |-> 2, b |-> "B" ] IN
+  /\ m.a > 1        \* type OK
+  /\ m.b = "B"      \* type OK
+  /\ m.c = { 1, 2 } \* should flag a type error
+```
+
+Moreover, the following example would require a type annotation:
+
+```tla
+RowAccess(m) ==
+    m.a > 0         \* should flag a type error
+```
+
+In the above example, the type checker would not be able to infer the type of
+`m` and would require an explicit type annotation:
+
+
+```tla
+\* @type: [ a: Int, b: Str ];
+RowAccess(m) ==
+    m.a > 0         \* should flag a type error
+```
+
+**Blocker.** It is not clear to me, what type we would assign to the record
+access operator. Although the type of `m` in `FieldAccess` is obvious to a
+human reader, as it simply requires us to do the top-to-the-bottom type
+propagation, Snowcat constructs a set of equality constraints that are solved
+by unification. This approach requires that we capture the record access
+operator `.` as an equality over type variables and types.
+
+### 4.3. Support plain records and variants, including row typing
+
+This solution follows the approach outlined by [Leijen05][], except that we
+require field names to be unique, not scoped. We present our approach in the
+next section.
+
+## 5. Solution
+
+In the following, we present row types as type terms. We discuss the
+user-facing syntax of the type system later in the text.
+
+### 5.1. Plain records
+
+By using [Row types][], we should be able to infer the type of `m` in
+`RowAccess`:
+
+```
+Rec(RowCons("a", Int, z))
+```
+
+In this example, `RowCons("a", Int, z)` is a type term indicating that the type
+has the field called `a`, and its type is `Int`. On top of that, this row
+extends a parametric type `z`, which either contains a non-empty sequence of
+rows, or is an empty sequence, that is, `RowNil`. Importantly, `RecCons` is
+wrapped with the term `Rec`, so no additional fields can be added to the type.
```

Comment:
This feels like a lot of info crammed into 1 paragraph. I'd suggest making separate, detailed entries for both operators. It's easy to miss the purpose of `Rec`, in that it acts as a guarantee of the absence of unspecified fields.

Reply:
Once this ADR is implemented, we have to extend the memo on type inference with more details. It's hard to add more details without writing a full paper.